src/Package/Impl/DataInspect/Office/CsvAppFileIO.cs

```diff
@@ -0,0 +1,108 @@
+            var variableName = index >= 0 ? expression.Substring(index + VariableProvider.GlobalEnvironmentExpression.Length + 1) : expression;
+
+            var csvFileName = MakeCsvFileName(variableName);
+            var file = ProjectUtilities.GetUniqueFileName(folder, csvFileName, "csv", appendUnderscore: true);
+            var rfile = file.Replace('\\', '/');
+
+            string currentStatusText = string.Empty;
+            await VsAppShell.Current.DispatchOnMainThreadAsync(() => {
+                var statusBar = VsAppShell.Current.GetGlobalService<IVsStatusbar>(typeof(SVsStatusbar));
+                statusBar.GetText(out currentStatusText);
+            });
+
+            try {
+                await SetStatusTextAsync(Resources.Status_WritingCSV);
+                using (var e = await session.BeginInteractionAsync()) {
+                    await e.RespondAsync(Invariant($"write.csv({variableName}, file='{rfile}')") + Environment.NewLine);
+                }
+
+                Task.Run(() => {
+                    try {
+                        Process.Start(file);
+                    } catch(Win32Exception ex) {
+                        VsAppShell.Current.ShowErrorMessage(
+                            string.Format(CultureInfo.InvariantCulture, Resources.Error_CannotOpenCsv, ex.Message));
+                    }
+                }).DoNotWait();
+            } finally {
+                await SetStatusTextAsync(currentStatusText);
+            }
+
+            Interlocked.Exchange(ref _busy, 0);
+        }
+
+        private static async Task SetStatusTextAsync(string text) {
+            await VsAppShell.Current.DispatchOnMainThreadAsync(() => {
+                var statusBar = VsAppShell.Current.GetGlobalService<IVsStatusbar>(typeof(SVsStatusbar));
+                statusBar.SetText(text);
+            });
+        }
+
+        private static string MakeCsvFileName(string variableName) {
+            var project = ProjectUtilities.GetActiveProject();
+            var projectName = project?.FileName;
+
+            var contentTypeService = VsAppShell.Current.ExportProvider.GetExportedValue<IContentTypeRegistryService>();
+            var viewTracker = VsAppShell.Current.ExportProvider.GetExportedValue<IActiveWpfTextViewTracker>();
+
+            var activeView = viewTracker.GetLastActiveTextView(contentTypeService.GetContentType(RContentTypeDefinition.ContentType));
+            var filePath = activeView.GetFilePath();
+
+            var csvFileName = string.Empty;
+
+            if (!string.IsNullOrEmpty(projectName)) {
+                csvFileName += Path.GetFileNameWithoutExtension(projectName);
+                csvFileName += "_";
+            }
+
+            if (!string.IsNullOrEmpty(filePath)) {
+                csvFileName += Path.GetFileNameWithoutExtension(filePath);
+                csvFileName += "_";
+            }
+
+            csvFileName += variableName.Replace('.', '_').Replace('@', '_').Replace('$', '_');
```

Comment:
Same comment regarding arbitrary expressions applies here.

However, even considering just the existing implementation, an expression of the form `foo[[1]]$baz` or `foo[['bar']]$baz` is also possible, given something like:
```r
e <- as.environment(list(baz = as.data.frame(0)))
foo <- list(e, bar = e)
```
The latter is particularly problematic, since it has a string literal that can contain pretty much anything.